File: 1918-maximum-score-of-a-good-subarray/1918-maximum-score-of-a-good-subarray.py
from typing import List

class Solution:
    def maximumScore(self, nums: List[int], k: int) -> int:
        N = len(nums)
        lo = k
        hi = k
        ans = nums[k]
        mn = nums[k]

        while lo >= 0 and hi < N:
            ans = max(ans, mn * (hi - lo + 1))
            loMn = min(mn, nums[lo - 1] if lo - 1 >= 0 else 0)
            hiMn = min(mn, nums[hi + 1] if hi + 1 < N else 0)

            if loMn >= hiMn:
                mn = loMn
                lo -= 1
            else:
                mn = hiMn
                hi += 1

        return ans

File: 1918-maximum-score-of-a-good-subarray/1918-maximum-score-of-a-good-subarray.cpp
class Solution {
public:
    int maximumScore(vector<int>& nums, int k) {
        const int n = nums.size();
        int maxScore = nums[k];
        int minVal = nums[k];
        int left = k;
        int right = k;

        while (left > 0 || right < n - 1) {
            if (left == 0)
                ++right;
            else if (right == n - 1)
                --left;
            else if (nums[left - 1] < nums[right + 1])
                ++right;
            else
                --left;
            minVal = min({minVal, nums[left], nums[right]});
            maxScore = max(maxScore, minVal * (right - left + 1));
        }

        return maxScore;
    }
};
